hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateMap.java

```diff
@@ -272,162 +178,145 @@ List<Pipeline> getPipelines(ReplicationConfig replicationConfig,
     Preconditions
         .checkNotNull(excludePipelines, "Pipeline exclude list cannot be null");
 
-    List<Pipeline> pipelines = null;
-    if (state == PipelineState.OPEN) {
-      pipelines = new ArrayList<>(query2OpenPipelines.getOrDefault(
-          replicationConfig, Collections.emptyList()));
-      if (excludeDns.isEmpty() && excludePipelines.isEmpty()) {
-        return pipelines;
-      }
-    } else {
-      pipelines = new ArrayList<>(pipelineMap.values());
-    }
-
-    Iterator<Pipeline> iter = pipelines.iterator();
-    while (iter.hasNext()) {
-      Pipeline pipeline = iter.next();
-      if (!pipeline.getReplicationConfig().equals(replicationConfig) ||
-          pipeline.getPipelineState() != state ||
-          excludePipelines.contains(pipeline.getId())) {
-        iter.remove();
-      } else {
-        for (DatanodeDetails dn : pipeline.getNodes()) {
-          if (excludeDns.contains(dn)) {
-            iter.remove();
-            break;
-          }
-        }
-      }
-    }
-
-    return pipelines;
+    return getPipelines(pipeline -> pipeline.getPipelineState() == state &&
```

Comment:
I can see a problem here, as the old code handled open pipelines in a special case, where all the open pipelines for a given replication config are stored in a list, and that list can be returned directly.

The new code must iterate all pipelines to find a list of those to return so it will be much slower than the original.

The reason open pipelines are managed this way, is because that is a common request from other parts of the system. Either to get the list of open pipelines or to get the count.

For example the count of open EC-6-3 pipelines can be answered with a simple map.get().size() call, with no iteration or allocation of any objects, but this change would break that fast lookup.